xbmc/windows/GUIMediaWindow.cpp

```diff
@@ -1488,19 +1488,170 @@ bool CGUIMediaWindow::OnPlayMedia(int iItem, const std::string &player)
  *
  * This function is called by OnClick()
  */
+class QueueMediaAsync : public CJob
+{
+public:
+  QueueMediaAsync(int iPlaylist, std::unique_ptr<CFileItemList> Localm_vecItems)
+  {
+    m_iPlaylist = iPlaylist;
+    m_Localm_vecItems.reset(new CFileItemList);
+    m_Localm_vecItems->Copy(*Localm_vecItems);
+  }
+
+  ~QueueMediaAsync(void) {}
+
+  bool DoWork(void)
+  {
+    bool ItemDetected = false;
+    int PlaylistSize = 0;
+    int Localm_vecItemsSize = m_Localm_vecItems->Size();
+
+    for ( int i = 0; i < Localm_vecItemsSize; i++ )
+    {
+      if (ShouldCancel(0, 0))
+        return false;
+
+      // Perform Playlist refresh every 50 items
+      if ( i % 50 == 0 )
+      {
+        CGUIMessage msg(GUI_MSG_PLAYLIST_CHANGED, 0, 0);
+        CServiceBroker::GetGUI()->GetWindowManager().SendMessage(msg);
+        Event.WaitMSec(100);
+      }
+
+      CFileItemPtr nItem = m_Localm_vecItems->Get(i);
+      PlaylistSize = CServiceBroker::GetPlaylistPlayer().GetPlaylist(m_iPlaylist).size();
+
+      //Verify if already added in "OnPlayAndQueueMedia". If that's the case, it must be at the end of the playlist, so only verify from index i to PlaylistSize
+      for ( int j = i; j < PlaylistSize; j++ )
+      {
+        CFileItemPtr pItem = CServiceBroker::GetPlaylistPlayer().GetPlaylist(m_iPlaylist).GetItem(j);
+
```

Comment:
const CFile...

No need to obtain playlist on each iteration. Do this once outside the loop.